### PR TITLE
TorchScript graph specialization on autocast state

### PR DESCRIPTION
**Pull request by tlemo**
This PR enables mixing eager mode and scripted autocast, for example:

```python
        @torch.jit.script
        def fn(a, b):
            return torch.mm(a, b)

        with autocast(enabled=True):
            result = fn(self.a_fp32, self.b_fp32)
```

This may dramatically improve the usability of scripted autocast, for example, looking at the official [AMP example][6]:

```python
for epoch in range(epochs):
    for input, target in zip(data, targets):
        with torch.cuda.amp.autocast(enabled=use_amp):
            output = net(input)
            loss = loss_fn(output, target)
        ...
```

A reasonable expectation might be to substitute `net` with the scripted version:

```python
net_jit = torch.jit.script(net)
...
for epoch in range(epochs):
    for input, target in zip(data, targets):
        with torch.cuda.amp.autocast(enabled=use_amp):
            output = net_jit(input) # this will not work
            loss = loss_fn(output, target)
       ...
```

The approach used here is to specialize at the `GraphFunction` level, using the autocast state (`at::autocast::is_enabled()` as specialization key). The main alternative would be using a custom IR node similar to `prim::RequiresGradCheck` - which should work, although the changes are likely more complex and more importantly it would force us to specialize the graphs upfront (at least the "hot/default" path).

Specializing at `GraphFunction` level allows a completely lazy specialization, which means that if there's only one autocast state used (which is most likely the case in real world usage), then we only need to specialize the for case that's actually used.

### CODE REVIEW

File: torch/csrc/jit/api/function_impl.h [hunk 1]
```diff
@@ -105,23 +108,35 @@ struct TORCH_API GraphFunction : public Function {
...
     }
     check_single_output();
-    executor_ = GraphExecutor(optimized_graph(), name_.name());
-    return executor_;
+    executor = GraphExecutor(optimized_graph(), name_.name());
+    return executor;
   }
 
+ private:
+  enum SpecializationKey {
```
Comment (jjsjann123) [inline torch/csrc/jit/api/function_impl.h:121]: SpecializationKey as an enum means specializations are exclusive to each other?

**Review by jjsjann123 — COMMENTED** [review]
My main concern here is the graph specialization design. Having an exclusive specialization key is quite restricting.
If our target is just to have something specialized for amp then composability is not relevant here. Nevertheless, having a narrow specialization inside function_impl seems quite intrusive.

**Comment by jjsjann123** [comment]
> the advantage of specializing inside `GraphExecutor` is that a specialized graph will be encapsulated as part of `GraphExecutor` state

Have we started work on graph specialization in `GraphExecutor`? If this is an open topic maybe we can insert us there instead.

**Comment by tlemo** [comment]
> My main concern here is the graph specialization design. Having an exclusive specialization key is quite restricting.

You mean we can't specialize on multiple dimensions? This is true, although that's not a fundamental part of the design - it's just that today we don't have any other use cases for the specialization. If anything comes up, changing the key to be a hash of multiple specialization axes is a trivial change.

**Comment by tlemo** [comment]
> > the advantage of specializing inside `GraphExecutor` is that a specialized graph will be encapsulated as part of `GraphExecutor` state
> 
> Have we started work on graph specialization in `GraphExecutor`? If this is an open topic maybe we can insert us there instead.

We discussed it, although it raises the same concerns and it's a bit more complex to implement (IMO the extra complexity could make the specialization more fragile - the nice thing about specializing stuff at GraphFunction level is that there's a clear barrier between specializations)

**Review by jjsjann123 — APPROVED** [review]
Stamping so we can move forward with amp support in our branch
